gnd/src/main/java/com/google/android/gnd/repository/FeatureRepository.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.repository;
+
+import com.google.android.gnd.model.Mutation;
+import com.google.android.gnd.model.Project;
+import com.google.android.gnd.model.feature.Feature;
+import com.google.android.gnd.model.feature.FeatureMutation;
+import com.google.android.gnd.persistence.local.LocalDataStore;
+import com.google.android.gnd.persistence.remote.RemoteDataEvent;
+import com.google.android.gnd.persistence.remote.RemoteDataStore;
+import com.google.android.gnd.persistence.sync.DataSyncWorkManager;
+import com.google.common.collect.ImmutableSet;
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import io.reactivex.Maybe;
+import java8.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Coordinates persistence and retrieval of {@link Feature} instances from remote, local, and in
+ * memory data stores. For more details on this pattern and overall architecture, see
+ * https://developer.android.com/jetpack/docs/guide.
+ */
+@Singleton
+public class FeatureRepository {
+  private final LocalDataStore localDataStore;
+  private final ProjectRepository projectRepository;
+  private final DataSyncWorkManager dataSyncWorkManager;
+
+  @Inject
+  public FeatureRepository(
+      LocalDataStore localDataStore,
+      RemoteDataStore remoteDataStore,
+      ProjectRepository projectRepository,
+      DataSyncWorkManager dataSyncWorkManager) {
+    this.localDataStore = localDataStore;
+    this.projectRepository = projectRepository;
+    this.dataSyncWorkManager = dataSyncWorkManager;
+
+    streamFeaturesToLocalDb(remoteDataStore);
+  }
+
+  /**
+   * Mirrors features in the current project from the remote db into the local db when the network
+   * is available. When invoked, will first attempt to resync all features from the remote db,
+   * subsequently syncing only remote changes.
+   */
+  private void streamFeaturesToLocalDb(RemoteDataStore remoteDataStore) {
```

Comment:
N.B. Most of the following doesn't apply if we decouple features from projects (as noted in the comments in this file).

I wonder a bit about having methods that rely internally on the current project and methods that take a project as argument. 

It's more verbose, but I think it simplifies things if we stick to a standard functional signature that expresses the relational constraints between the entities. e.g. all features require a project context so all feature functions are mappings from : project -> feature(s). 

This makes functions like this one slightly more self-explanatory. As this function currently stands, one needs to read the impl or the doc to know that it uses the current project context. 

Taking the project context explicitly in each method also makes these methods composable at the point of use and more flexible e.g. 

```java
projectRepo.getActiveProjectOnceAndStream().switchMap(featureRepo::doSomethingWithFeatures).switchMap(datastore::DosomethingWithDataStore).subscribe()
```

This keeps our APIs functional and usable as component-wise compositions, rather than surfacing preconceived uses/sequences as a single method. Our api methods are the morphemes (smallest possible composable units) and rx is the grammar.

This approach should also give callers control over subscriptions (partially tackling the problem noted in the TODO on this method, from the API side, we have no idea when the disposable might be garbage collected, but if we pass over subscription control to the caller, it's more likely they'll be able to anticipate and account for possible GC calls).

Reply:
Valid observation and good suggestions for improvement! Let's follow up on this when we fix the poorly defined subscription pattern. Thanks!